contracts/uniswap/UniswapReserve.sol

```diff
@@ -0,0 +1,300 @@
+    // Parts per 10000
+    uint public constant DEFAULT_FEE_BPS = 25;
+
+    UniswapFactory public uniswapFactory;
+    address public kyberNetwork;
+
+    uint public feeBps = DEFAULT_FEE_BPS;
+
+    // Uniswap exchange contract for every listed token
+    // token -> exchange
+    mapping (address => address) public tokenExchange;
+
+    bool public tradeEnabled = true;
+
+    /**
+        Constructor
+    */
+    function UniswapReserve(
+        UniswapFactory _uniswapFactory,
+        address _admin,
+        address _kyberNetwork
+    )
+        public
+    {
+        require(address(_uniswapFactory) != 0);
+        require(_admin != 0);
+        require(_kyberNetwork != 0);
+
+        uniswapFactory = _uniswapFactory;
+        admin = _admin;
+        kyberNetwork = _kyberNetwork;
+    }
+
+    function() public payable {
+        // anyone can deposit ether
+    }
+
+    /**
+        Returns dest quantity / source quantity.
+    */
+    function getConversionRate(
+        ERC20 src,
+        ERC20 dest,
+        uint srcQty,
+        uint blockNumber
+    )
+        public
+        view
+        returns(uint)
+    {
+        // This makes the UNUSED warning go away.
+        blockNumber;
+
+        require(isValidTokens(src, dest));
+
+        if (!tradeEnabled) return 0;
+
+        ERC20 token;
+        if (src == ETH_TOKEN_ADDRESS) {
+            token = dest;
+        } else if (dest == ETH_TOKEN_ADDRESS) {
+            token = src;
+        } else {
+            // Should never arrive here - isValidTokens requires one side to be ETH
+            revert();
+        }
+
+        UniswapExchange exchange = UniswapExchange(tokenExchange[token]);
+
+        uint convertedQuantity;
+        if (src == ETH_TOKEN_ADDRESS) {
+            uint quantity = srcQty * (10000 - feeBps) / 10000;
+            convertedQuantity = exchange.getEthToTokenInputPrice(quantity);
+        } else {
+            convertedQuantity = exchange.getTokenToEthInputPrice(srcQty);
+            convertedQuantity = convertedQuantity * (10000 - feeBps) / 10000;
+        }
+
+        return calcRateFromQty(
+            srcQty, /* srcAmount */
+            convertedQuantity, /* destAmount */
+            getDecimals(src), /* srcDecimals */
+            getDecimals(dest) /* dstDecimals */
+        );
+    }
+
+    event TradeExecute(
+        address indexed sender,
+        address src,
+        uint srcAmount,
+        address destToken,
+        uint destAmount,
+        address destAddress
+    );
+
+    /**
+      conversionRate: expected conversion rate should be >= this value.
+     */
+    function trade(
+        ERC20 srcToken,
+        uint srcAmount,
+        ERC20 destToken,
+        address destAddress,
+        uint conversionRate,
+        bool validate
+    )
+        public
+        payable
+        returns(bool)
+    {
+        // Not using this variable that is part of the interface.
```

Comment:
there is a general question here.
is there an attack scenario where destAmount can change between getRate and trade?
this can happen due to any external contract call between the two.
was this option tested?

Reply:
in general the easiest defense for this is to verify rate is at least conversionRate in  parameter.